### ISSUE TITLE
zmq.tests.test_device.TestDevice.test_single_socket_forwarder_bind() and zmq.tests.test_device.TestDevice.test_single_socket_forwarder_connect(): Failures and errors

### ISSUE THREAD

**Issue by Arfrever** [zeromq/pyzmq#681]
`zmq.tests.test_device.TestDevice.test_single_socket_forwarder_bind()` and `zmq.tests.test_device.TestDevice.test_single_socket_forwarder_connect()` cause failures and errors with all versions of Python (at least CPython 2.6, 2.7, 3.2, 3.3, 3.4) in both PyZMQ 14.6.0 and git trunk.
I use ZeroMQ 4.1.1.

Result of test suite of PyZMQ git trunk with CPython 3.4:
```
$ PYTHONPATH="build/lib.linux-x86_64-3.4" nosetests-3.4 -s -v -w build/lib.linux-x86_64-3.4/zmq/tests
...
test_device_attributes (zmq.tests.test_device.TestDevice) ... ok
test_device_types (zmq.tests.test_device.TestDevice) ... ok
test_proxy (zmq.tests.test_device.TestDevice) ... ok
test_single_socket_forwarder_bind (zmq.tests.test_device.TestDevice) ... FAIL
ERROR
test_single_socket_forwarder_connect (zmq.tests.test_device.TestDevice) ... FAIL
ERROR
test_tsdevice_attributes (zmq.tests.test_device.TestDevice) ... ok
test_again (zmq.tests.test_error.TestZMQError) ... ok
test that strerror gets the right type. ... ok
...
======================================================================
ERROR: test_single_socket_forwarder_bind (zmq.tests.test_device.TestDevice)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/pyzmq/build/lib.linux-x86_64-3.4/zmq/tests/__init__.py", line 103, in tearDown
    raise RuntimeError("context could not terminate, open sockets likely remain in test")
RuntimeError: context could not terminate, open sockets likely remain in test

======================================================================
ERROR: test_single_socket_forwarder_connect (zmq.tests.test_device.TestDevice)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/pyzmq/build/lib.linux-x86_64-3.4/zmq/tests/__init__.py", line 103, in tearDown
    raise RuntimeError("context could not terminate, open sockets likely remain in test")
RuntimeError: context could not terminate, open sockets likely remain in test

======================================================================
FAIL: test_single_socket_forwarder_bind (zmq.tests.test_device.TestDevice)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/pyzmq/build/lib.linux-x86_64-3.4/zmq/tests/test_device.py", line 78, in test_single_socket_forwarder_bind
    self.assertEqual(msg, self.recv(req))
  File "/tmp/pyzmq/build/lib.linux-x86_64-3.4/zmq/tests/__init__.py", line 165, in recv
    return self._select_recv(False, socket, **kwargs)
  File "/tmp/pyzmq/build/lib.linux-x86_64-3.4/zmq/tests/__init__.py", line 157, in _select_recv
    assert len(r) > 0, "Should have received a message"
AssertionError: Should have received a message

======================================================================
FAIL: test_single_socket_forwarder_connect (zmq.tests.test_device.TestDevice)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/pyzmq/build/lib.linux-x86_64-3.4/zmq/tests/test_device.py", line 49, in test_single_socket_forwarder_connect
    self.assertEqual(msg, self.recv(req))
  File "/tmp/pyzmq/build/lib.linux-x86_64-3.4/zmq/tests/__init__.py", line 165, in recv
    return self._select_recv(False, socket, **kwargs)
  File "/tmp/pyzmq/build/lib.linux-x86_64-3.4/zmq/tests/__init__.py", line 157, in _select_recv
    assert len(r) > 0, "Should have received a message"
AssertionError: Should have received a message

----------------------------------------------------------------------
Ran 165 tests in 30.327s

FAILED (SKIP=18, errors=2, failures=2)
```

**Comment by minrk** [comment]
This is due to a breaking, unintentional change in libzmq-4.0.6 and 4.1.1. I'm working with libzmq folks to determine whether the change should be reverted before I change this test.

**Comment by minrk** [comment]
The upstream bug is fixed in libzmq master, and will be fixed in 4.0.7 and 4.1.2.
